### DISCUSSION TITLE
Links in emails in reverse proxy setup

### DISCUSSION THREAD

**Discussion by ocanam** [kimai/kimai#3182] [Questions & Answers] [answered]
**Version Info**
 - Kimai version: 1.14.1
 - PHP version: 7.3.27-1~deb10u1
 - Server: Debian 10 (Linux 4.19.0-16-amd64)

**Setup**
Classic reverse proxy situation
`Kimai2 (http://timesheet-backend.fake.domain:8080) <-> Apache2 SSL RProxy (https://timesheet.fake.domain) <-> Client`

**Problem**
The emails sent by kimai can contain http links that point to the Kimai2 host, rendering those links unusable as my backend server can't be reached directly from outside.
So
What I have in the emails is:
http://timesheet-backend.fake.domain:8080/index.php/en/resetting/reset/blah
What I need is
https://timesheet.fake.domain/index.php/en/resetting/reset/blah

Not sure it's a bug, a missing feature or a misconfiguration on my side.
Whatever category it is, I didn't found where I could configure the wanted behavior for links in email.
Any help or suggestion are welcome.

Thanks

**Comment by kevinpapst** [comment]
Hm, interesting question. 
I think this should be handled transparently without further configurations in Kimai itself. And Kimai is used frequently behind proxies, so I assume that should have received that feedback already if there is a general problem.

Short answer: I am not sure ... so let me start with some questions:

Everything else works?
Did you read and configure accordingly:
 https://www.kimai.org/documentation/webserver-configuration.html#reverse-proxy ?

The links in Kimai web are all relative/absolute but without hostname/port ?

**Comment by ocanam** [comment]
Thank for the feedback.

To answer your questions:

yes everything else works fine.
`TRUSTED_PROXIES` is set in the .env according to the setup.
So in my environment
`TRUSTED_PROXIES=timesheet.fake.domain`

As indicated in the initial message, the links in the emails are absolute ones, with the protocol, host and port of the kimai host.

Not sure how it could be handled transparently, because the emails don't go through the web proxy, so there is no possibilities to rewrite the urls.
And there is no http header to inform the kimai server about its proxy context when generating and sending an email.

I know in other applications we run, redmine for example, the base url for emails must be configured manually.

**Comment by kevinpapst** [comment]
The only emails that exist right now are the ones that are triggered by the user directly (register, password forgotten) and they are rendered in web context. So I assumed that the router knows the context, but maybe not.

That is what I found: https://symfony.com/doc/current/routing.html#generating-urls-in-commands

Try to add that to config/packages/local.yaml (create file if it does not exist):
```
framework:
    router:
        default_uri: 'https://timesheet.fake.domain/'
```
and then run `bin/console kimai:reload --env=prod`

**Comment by ocanam** [comment]
Symfony did not appreciate
```
[ERROR] Unrecognized option "default_uri" under "framework.router". Available options are "enabled", "http_port",
         "https_port", "resource", "strict_requirements", "type", "utf8".
```
Regarding the Symfony version, does using kimai2 v1.14.1 change something compared to the latest release v1.14.3 ? (don't think so)

> 
> The only emails that exist right now are the ones that are triggered by the user directly (register, password forgotten) and they are rendered in web context. So I assumed that the router knows the context, but maybe not.
> 
Assuming there is a workaround via the Symfony router, IMHO the day you want kimai to send spontaneous emails such as notifications or reminders, not related to a user activity via the browser, and so not related to a web context, it won't work anymore.

**Comment by kevinpapst** [comment]
I am sorry, copied URL to the wrong Symfony version:
https://symfony.com/doc/4.4/routing.html#generating-urls-in-commands

Still in local.yaml:
```
parameters:
    router.request_context.host: 'example.org'
    router.request_context.base_url: 'my/path'
    asset.request_context.base_path: '%router.request_context.base_url%'
```

Please share your config if you get it up and running! 
I will create a documentation page and maybe add a custom setting for Kimai for new emails in the future.

**Comment by ocanam** [comment]
Nope :(
This time it didn't prevent symfony to reload properly but it didn't change anything either to the links in the emails.

**Comment by kevinpapst** [comment]
Hm, thats a pity. In that case someone needs to dig deeper.
Maybe trusted proxies is wrong, maybe something else. 

The complete security code for login & password reset has changed for the next version, so I am not going to debug it anymore. But if you want to give master branch a try, you could start debugging here: https://github.com/kevinpapst/kimai2/blob/master/src/Controller/Security/PasswordResetController.php#L180

**Comment by Silvoax** [comment, marked as answer]
Hello,

Just to let you know, I had exactly the same problem with a kimai2 application running on a docker behind a proxy (Apache). URL in email for resetting passwords was the local url and not the public one.
If your proxy is based on Apache with proxy reverse instruction you can follow instructions here : https://serverfault.com/questions/557233/apache-proxy-http-redirect-to-ip-and-set-hostname
It worked for me.

TLDR : Just add "ProxyPreserveHost On" instruction just after your proxy reverse instruction. I think there is a same kind of instruction on any proxy system, to send the initial hostname to the application behind the proxy.

I hope it will be helpfull.

**Reply by kevinpapst** [reply]
Thanks for sharing your solution. I converted the issue to a discussion, to be able to select yours as answer 👍
